### ISSUE TITLE
`gazelle:default_visibility` directive does not set visibility of library

### ISSUE THREAD

**Issue by LINKIWI** [bazelbuild/bazel-gazelle#1638]
### What version of gazelle are you using?

v0.33.0

### What version of rules_go are you using?

v0.41.0

### What version of Bazel are you using?

6.3.2

### Does this issue reproduce with the latest releases of all the above?

Yes

### What operating system and processor architecture are you using?

`Linux 6.2.0-32-generic #32~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Aug 18 10:40:13 UTC 2 x86_64 x86_64 x86_64 GNU/Linux`

### What did you do?

I want to force all Gazelle-generated targets for a `go_repository` to be public. `gazelle:default_visibility` claims to do this; it does not.

Minimal reproducible example is as follows:

`WORKSPACE`:

```starlark
workspace(name = "example")

load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

http_archive(
    name = "io_bazel_rules_go",
    sha256 = "278b7ff5a826f3dc10f04feaf0b70d48b68748ccd512d7f98bf442077f043fe3",
    urls = ["https://github.com/bazelbuild/rules_go/releases/download/v0.41.0/rules_go-v0.41.0.zip"],
)

load("@io_bazel_rules_go//go:deps.bzl", "go_register_toolchains", "go_rules_dependencies")

go_rules_dependencies()

go_register_toolchains(version = "1.21.1")

http_archive(
    name = "bazel_gazelle",
    sha256 = "d3fa66a39028e97d76f9e2db8f1b0c11c099e8e01bf363a923074784e451f809",
    urls = ["https://github.com/bazelbuild/bazel-gazelle/releases/download/v0.33.0/bazel-gazelle-v0.33.0.tar.gz"],
)

load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies", "go_repository")

gazelle_dependencies()

go_repository(
    name = "com_github_yuin_goldmark",
    importpath = "github.com/yuin/goldmark",
    sum = "h1:fVcFKWvrslecOb/tg+Cc05dkeYx540o0FuFt3nUVDoE=",
    version = "v1.4.13",
)

go_repository(
    name = "org_golang_x_mod",
    importpath = "golang.org/x/mod",
    sum = "h1:rmsUpXtvNzj340zd98LZ4KntptpfRHwpFOHG188oHXc=",
    version = "v0.12.0",
)

go_repository(
    name = "org_golang_x_net",
    importpath = "golang.org/x/net",
    sum = "h1:ugBLEUaxABaB5AJqW9enI0ACdci2RUd4eP51NTBvuJ8=",
    version = "v0.15.0",
)

go_repository(
    name = "org_golang_x_sync",
    importpath = "golang.org/x/sync",
    sum = "h1:ftCYgMx6zT/asHUrPw8BLLscYtGznsLAnjq5RH9P66E=",
    version = "v0.3.0",
)

go_repository(
    name = "org_golang_x_sys",
    importpath = "golang.org/x/sys",
    sum = "h1:CM0HF96J0hcLAwsHPJZjfdNzs0gftsLfgKt57wWHJ0o=",
    version = "v0.12.0",
)

go_repository(
    name = "org_golang_x_tools",
    build_directives = ["gazelle:default_visibility //visibility:public"],
    importpath = "golang.org/x/tools",
    sum = "h1:Iey4qkscZuv0VvIt8E0neZjtPVQFSc870HQ448QgEmQ=",
    version = "v0.13.0",
)
```

Note the key line above in the `go_repository` for `org_golang_x_tools`:

```
...
build_directives = ["gazelle:default_visibility //visibility:public"],
...
```

`BUILD`:

```starlark
load("@bazel_gazelle//:def.bzl", "gazelle")
load("@io_bazel_rules_go//go:def.bzl", "go_library")

# gazelle:prefix example
gazelle(name = "gazelle")

go_library(
    name = "example",
    embed = ["@org_golang_x_tools//cmd/file2fuzz:file2fuzz_lib"],
)
```

### What did you expect to see?

Expected behavior: `//:example` compiles, because `file2fuzz_lib` is declared public.

### What did you see instead?

Actual behavior: compile failure:

```
target '@org_golang_x_tools//cmd/file2fuzz:file2fuzz_lib' is not visible from target '//:example'. Check the visibility declaration of the former target if you think the dependency is legitimate
```

If I inspect the generated build file, the library is still set to private visibility.

```starlark
load("@io_bazel_rules_go//go:def.bzl", "go_binary", "go_library", "go_test")

go_library(
    name = "file2fuzz_lib",
    srcs = ["main.go"],
    importpath = "golang.org/x/tools/cmd/file2fuzz",
    visibility = ["//visibility:private"],
)

go_binary(
    name = "file2fuzz",
    embed = [":file2fuzz_lib"],
    visibility = ["//visibility:public"],
)

go_test(
    name = "file2fuzz_test",
    srcs = ["main_test.go"],
    embed = [":file2fuzz_lib"],
)
```

How do I *force* Gazelle to make this target public visibility?

**Comment by hunshcn** [comment]
same issue
This makes it difficult to add x_defs.

@fmeum Is this the expected performance?

**Comment by fmeum** [comment]
I would say no, let's change the behavior if the user explicitly requests it in this way.

@linzhp Any concerns?

**Comment by linzhp** [comment]
I am fine with that. I am also not happy with `go_visibility`'s current behavior: it only appends additional visibility to internal packages. I would like it to override the visibility, e.g., to stop a deprecated 3rd party library from being used by new projects. 

So I kind of regret to call it `go_visibility` now. It should have been `go_add_visilbility` or `go_extra_visibility` (https://github.com/bazelbuild/bazel-gazelle/pull/649#pullrequestreview-295059343), so we can have a `go_override_visibility`.

**Comment by dcrosta** [comment]
Will the fix in #1794 also make the `default_visibility` apply to `go_test` targets?

Come to think of it, should this directive cause gazelle to emit a `package(default_visibility=[<whatever>])` line instead of setting it per-target?

**Comment by hunshcn** [comment]
> I am fine with that. I am also not happy with `go_visibility`'s current behavior: it only appends additional visibility to internal packages. I would like it to override the visibility, e.g., to stop a deprecated 3rd party library from being used by new projects.
> 
> So I kind of regret to call it `go_visibility` now. It should have been `go_add_visilbility` or `go_extra_visibility` ([#649 (review)](https://github.com/bazelbuild/bazel-gazelle/pull/649#pullrequestreview-295059343)), so we can have a `go_override_visibility`.

Maybe it can be divided into two steps to do.
1. At present, the responsibility of `go_visibility` is `add visibility`, but it cannot affect go_binary. This is a semantic inconsistency, and we need to make it effective.
2. new `go_override_visibility`. All the visibility processed by `go_visibility` can be overwritten by `go_override_visibility`.

#1794 do the first one
